### ISSUE TITLE
Opportunity for optimization

### ISSUE THREAD

**Issue by lambertlb** [icsharpcode/ILSpy#1689]
I am going through C# spec and adding tests and tried "Null-conditional operators ?. and ?[]". The code translates and runs just fine but is a lot larger than the original leaving an opportunity for some pattern matching and substitution. Since the code works i would set this as low priority.

Original code:
```csharp
private void ByteMemberAccessTests()
{
	Byte[] array1 = new Byte[] { 1, 2, 3, 4, 5 };
	Program.LocalAssert(array1[2] == 3);
	Dictionary<Byte, String> dict = new Dictionary<Byte, String>();
	dict[1] = "one";
	dict[2] = "two";
	Program.LocalAssert(dict[1] == "one");
	Program.LocalAssert(dict[2] == "two");
	List<Byte[]> setsOfNumbers = null;
	Program.LocalAssert(setsOfNumbers?[0]?.ToString() == null);
	setsOfNumbers = new List<byte[]>()
	{
		new Byte[] { 1, 2, 3 },
		null
	};
	Program.LocalAssert(setsOfNumbers?[0]?[1].ToString() == "2");
	Program.LocalAssert(setsOfNumbers?[1]?[1].ToString() == null);
}
```

Translated code:
```csharp
private void ByteMemberAccessTests()
{
	byte[] array = new byte[5]
	{
		1,
		2,
		3,
		4,
		5
	};
	Program.LocalAssert(array[2] == 3);
	Dictionary<byte, string> dict = new Dictionary<byte, string>();
	dict[1] = "one";
	dict[2] = "two";
	Program.LocalAssert(dict[1] == "one");
	Program.LocalAssert(dict[2] == "two");
	List<byte[]> setsOfNumbers2 = null;
	object obj;
	if (setsOfNumbers2 == null)
	{
		obj = null;
	}
	else
	{
		byte[] array2 = setsOfNumbers2[0];
		obj = ((array2 != null) ? array2.ToString() : null);
	}
	Program.LocalAssert(obj == null);
	List<byte[]> list = new List<byte[]>();
	list.Add(new byte[3]
	{
		1,
		2,
		3
	});
	list.Add(null);
	setsOfNumbers2 = list;
	object a;
	if (setsOfNumbers2 == null)
	{
		a = null;
	}
	else
	{
		byte[] array3 = setsOfNumbers2[0];
		a = ((array3 != null) ? array3[1].ToString() : null);
	}
	Program.LocalAssert((string)a == "2");
	object obj2;
	if (setsOfNumbers2 == null)
	{
		obj2 = null;
	}
	else
	{
		byte[] array4 = setsOfNumbers2[1];
		obj2 = ((array4 != null) ? array4[1].ToString() : null);
	}
	Program.LocalAssert(obj2 == null);
}
```

**Comment by lambertlb** [comment]
Every time i paste code into an issue the spaces go away. Is there a way to maintain the spaces?

**Comment by tamlin-mike** [comment]
> Is there a way to maintain the spaces?

Use markdown formatting. Three backtick's to start a code block, three to close. Link to help should be available at bottom (?) of page.

**Comment by dgrunwald** [comment]
The problem when applying `?[]` to `List<byte[]>` was caused by #1709, which is fixed now on the 5.0.x branch.

Applying `?[]` to arrays is a separate problem.
